src/udp/connection/encoded_id.rs

```diff
@@ -0,0 +1,346 @@
+            false => LocalClock::next_as_timestamp(&(lifetime + after_maybe)),
+        };
+
+        let mut hasher = DefaultHasher::default();
+        hasher.write_u64(now_maybe_next / lifetime.as_secs());
+        hasher.write(client_id.value().as_slice());
+
+        HashedEncodedId {
+            value: hasher.finish().to_le_bytes(),
+        }
+    }
+}
+
+impl EncodedId<KeyedClientId> for HashedEncodedId {
+    type Error = &'static str;
+
+    fn new(client_id: KeyedClientId, lifetime: Duration) -> Self {
+        Self::new_int(client_id, lifetime, false, None)
+    }
+
+    fn value(&self) -> &[u8; 8] {
+        &self.value
+    }
+
+    fn check(
+        client_id: KeyedClientId,
+        lifetime: Duration,
+        encoded_id: [u8; 8],
+        after_now: Option<Duration>,
+    ) -> Result<(), Self::Error> {
+        if Self::new_int(client_id.clone(), lifetime, false, after_now).value == encoded_id {
+            println!("Now");
+            return Ok(());
+        } else if Self::new_int(client_id.clone(), lifetime, true, after_now).value == encoded_id {
+            println!("Past");
+            return Ok(());
+        } else {
+            return Err("Expired connection id");
+        }
+    }
+}
+
+impl EncodedId<KeyedClientId> for WitnessEncodedId {
+    type Error = &'static str;
+
+    fn new(client_id: KeyedClientId, lifetime: Duration) -> Self {
+        let expiry_u32: u32 = LocalClock::next_as_timestamp(&lifetime).try_into().unwrap();
+
+        let mut hasher = DefaultHasher::default();
+        hasher.write_u32(expiry_u32);
+        hasher.write(client_id.value().as_slice());
+        let witness = hasher.finish().to_le_bytes();
+
+        let id: Vec<u8> = [&witness[0..4], expiry_u32.to_le_bytes().as_slice()].concat();
+
+        WitnessEncodedId {
+            value: id.try_into().unwrap(),
+        }
+    }
+
+    fn value(&self) -> &[u8; 8] {
+        &self.value
+    }
+
+    fn check(
+        client_id: KeyedClientId,
+        _lifetime: Duration,
+        encoded_id: [u8; 8],
+        after_now: Option<Duration>,
+    ) -> Result<(), Self::Error> {
+        let expiry_bytes: [u8; 4] = encoded_id[4..8].try_into().unwrap();
+        let expiry_u32 = u32::from_le_bytes(expiry_bytes);
+
+        let after_maybe = match after_now {
+            Some(after) => after,
+            None => Duration::ZERO,
+        };
+
+        let now_u32 = LocalClock::next_as_timestamp(&after_maybe).try_into().unwrap();
+
+        if expiry_u32 <= now_u32 {
+            return Err("Expired connection id");
+        }
+
+        let mut hasher = DefaultHasher::default();
+        hasher.write_u32(expiry_u32);
+        hasher.write(client_id.value().as_slice());
+
+        let witness = hasher.finish().to_le_bytes();
+
+        if &witness[0..4] != &encoded_id[0..4] {
+            Err("Bad Witness in Connection Id")
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl EncodedId<PlainClientId> for EncryptedEncodedId {
```

Comment:
@da2ce7 I would remove the other implementations if we finally only use the fastest version.